draft-ietf-quic-transport.md

```diff
@@ -1101,6 +1132,41 @@ suggested structure:
 -->
 
 
+## Required Operations on Connections
+
+There are certain operations which an application MUST be able to perform when
+interacting with the QUIC transport.  This document does not specify an API, but
+any implementation of this version of QUIC MUST expose the ability to perform
+the operations described in this section on a QUIC connection.
+
+When implementing the client role, applications need to be able to:
+
+- open a connection, which begins the exchange described in {{handshake}};
+- enable 0-RTT; and
+- be informed when 0-RTT has been accepted or rejected by a server.
+
+When implementing the server role, applications need to be able to:
+
+- listen for incoming connections, which prepares for the exchange described in
+  {{handshake}};
+- if Early Data is supported, embed application-controlled data in the TLS
+  resumption ticket sent to the client; and
+- if Early Data is supported, retrieve application-controlled data from the
+  client's resumption ticket and enable rejecting Early Data based on that
+  information.
```

Comment:
Servers need to expose whether incoming data is Early Data, correct? ie: for the Early Data header.

Reply:
Added to stream requirements, but now I'm not so sure that's correct.  Early-Data has to be added when the server "forwards a request prior to the completion of the TLS handshake".  That would mean the server really only needs a yes/no about the current state of the connection when the request is being forwarded.  If the handshake has completed, it doesn't matter that the request was originally early -- it's now retroactively validated as a non-replay by the successful handshake.

Reply:
Agreed, I think it's only yes/no about the current state of the connection.

Reply:
An app should also be able to reject incoming connections, for example based on implementation specific criteria such as load or peer IP address.

Reply:
That seems like a worthwhile feature for a QUIC implementation to expose, but doesn't seem like one we should mandate every QUIC implementation must expose.

Reply:
I can agree to that.